ncl/keymap-codegen.ncl

```diff
@@ -821,32 +821,47 @@ let validators = import "validators.ncl" in
     = {
       JsonValue = std.contract.from_validator json_value_validator,
 
-      # JSON serialization of key::chorded::Key is { chord: key, passthrough: key }
+      # JSON serialization of key::chorded::Key is { chords: [[id, key]], passthrough: key }
       json_value_validator =
         validators.record.validator {
-          fields_validator = validators.record.has_exact_fields ["chord", "passthrough"],
+          fields_validator = validators.record.has_exact_fields ["chords", "passthrough"],
           field_validators = {
-            chord = key.json_value_validator,
+            chords =
+              validators.array.validator (fun v =>
+                v
+                |> match {
+                  [id, ch_key] =>
+                    let valid_id = validators.is_number id in
+                    let valid_key = key.json_value_validator ch_key in
+                    [valid_id, valid_key]
+                    |> match {
+                      ['Ok, 'Ok] => 'Ok,
+                      [e, _] => e,
+                      [_, e] => e,
+                    },
+                  _ => 'Error { message = "expected [id, key]" },
+                }
+              ),
             passthrough = key.json_value_validator,
           },
         },
 
       is_json_value = fun k => 'Ok == json_value_validator k,
 
-      codegen_values = fun k @ { chord, passthrough } =>
-        let chord_cv = chord |> key.codegen_values |> chorded_nestable.wrap in
+      codegen_values = fun k @ { chords, passthrough } =>
+        let chords_cv = chords |> std.array.map (fun [id, ch_k] => [id, ch_k |> key.codegen_values |> chorded_nestable.wrap]) in
```

Comment:
`std.array.map (fun [id, ch_k] => [id, ch_k |> f])` is a pattern which occurs enough that it's worth factoring this out to a function.